refactor(lang): Update Compiler interface to produce a Program

wolffcm opened the pull request:
I did as little as possible to enable the vanilla Flux REPL to work.

Unit tests are not yet passing, I thought I would get feedback before I went to the trouble of fixing them to avoid unnecessary work.

I made `Controller.Query()` just trivially compile and run the `Program` produced by its `Compiler` argument.  As a result, `Controller.compileQuery` is totally unneeded now and caused compiler errors, so I just deleted its body.

Fixes #1086.

### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

nathanielc commented on lang/compiler.go:
```diff
@@ -79,3 +114,45 @@ func (ASTCompiler) CompilerType() flux.CompilerType {
+	}, nil
+}
+
+// TODO(cwolff): There should be an easy method presented by the plan package to do this.
```
Agreed can you create an issue to track fixing this? And link it to the parent #1085 issue so we track it as part of this whole process?

wolffcm replied:
I went ahead and did the work for this:
https://github.com/influxdata/flux/pull/1096
I was planning on merging it to master, and then rebasing this branch when it lands.

nathanielc commented on query.go:
```diff
@@ -6,10 +6,6 @@ import (
 
 // Query represents an active query.
 type Query interface {
```
I would have expected the Ready method to need to change to be a `<-chan Result` instead of a `<-chan map[string]Result`. The difference being that now the Ready channel can produce more than one set of results. In fact the name Ready is inaccurate now since its not about the query being ready but just a mechanism to deliver results.

If this should be a follow up PR that is fine, lets create an issue for it.

wolffcm commented:
@jsternberg 
> I like most of the PR, but I want to see if we can modify the interface so that the program potentially accepts an executor instead of creating its own.

@nathanielc mentioned that he doesn't think that `Start` should accept an `Executor` but I think that was more with respect to dependencies, rather than resource management.  It does seem like resource management needs to be handled somehow.

jsternberg commented:
Yea, I don't care too much about the execution itself, but just found executor to be a useful abstraction for that. Maybe we could make the dispatcher and memory allocator into interfaces in the flux package and pass those specifically? Or we can create a new `flux.ResourceManager` interface.

```
type ResourceManager interface {
    Allocator() *memory.Allocator // or the arrow allocator
    Dispatch(ctx context.Context, fn func()) // access to the dispatcher
}
```

wolffcm commented:
I pushed a commit that addresses feedback:
- `Program.Start` method now takes a `*memory.Allocator` as a parameter.  Nathaniel and I talked about making this more general but for now this will work.  
- I made `Query.Ready` return `chan Result`

nathanielc commented on query.go:
```diff
@@ -6,14 +6,10 @@ import (
 	// Its possible that the channel is closed before any results arrive,
 	// in which case the query should be inspected for an error using Err().
-	Ready() <-chan map[string]Result
+	Ready() <-chan Result
```
Can you rename this from `Ready` to `Results` or something similar? Ready implies that it fires once which is no longer the case.

nathanielc commented on result_iterator.go:
```diff
@@ -48,26 +48,32 @@ func (r *queryResultIterator) More() bool {
-	return r.results.Next()
+	if r.nextResult == nil {
+		r.nextResult = <-r.query.Ready()
+	}
```
Next is allowed to panic if More returns `false` and `Next` is called. That means that this check is not necessary and could possibly lead to odd bugs.

wolffcm replied:
I don't think that would quite get the behavior we want, but you're right there's logic errors here. I added a new commit (with comments) that addresses them.
